tests/framework/static_analysis.py

```diff
@@ -0,0 +1,596 @@
+        syscall instruction"""
+
+    @property
+    @abstractmethod
+    def syscall_argument_registers(self):
+        """The list of registers (in order) used to pass arguments to the architectures syscall instruction"""
+
+    @property
+    @abstractmethod
+    def fn_call_argument_registers(self):
+        """The list of registers (in order) used to pass arguments to normal function calls"""
+
+    @property
+    @abstractmethod
+    def seccomp_arch(self) -> seccomp.Arch:
+        """Convert to the correct variant of seccomp's Arch enum"""
+
+    @abstractmethod
+    def generalize_reg(self, reg: str) -> list[str]:
+        """For a given register, return a list of registers that partially alias it.
+
+        E.g. on x86, when given %rdi as input, return [%rdi, %edi, %di]"""
+
+    @abstractmethod
+    def is_call_instruction(self, insn: Instruction) -> bool:
+        """Checks whether the given instruction is a subroutine call"""
+
+    @abstractmethod
+    def is_syscall_instruction(self, insn: Instruction) -> bool:
+        """Checks whether the given instruction is a syscall instruction"""
+
+    @abstractmethod
+    def backpropagate_register(self, insn: Instruction, reg: str):
+        """
+        If this instruction loads an immediate into the given register, returns
+        that immediate as an integer. If the instruction is a register to register transfer,
+        returns the source register for this transfer. If this instruction doesn't change
+        the given register, returns the given register. Returns None if we don't know
+        how to backpropagate through this instruction.
+
+        :param insn: The instruction to backpropagate through
+        :param reg: the register to backpropagate through the given instruction
+        :return:
+        """
+
+    @abstractmethod
+    def instruction_from_str(self, insn_str: str) -> Instruction:
+        """Parses the given string as a single assembly instruction, in the syntax that
+        objdump uses by default on this architecture"""
+
+    def determine_register_value(self, instructions: list[Instruction], register: str):
+        """Determines the value of the given register at the end of the given instruction sequence
+        via backpropagation"""
+        looking_for = self.generalize_reg(register)
+        transforms = []
+
+        for insn in reversed(instructions):
+            for reg in looking_for:
+                next_reg = self.backpropagate_register(insn, reg)
+
+                if isinstance(next_reg, tuple):
+                    next_reg, transform = next_reg
+
+                    transforms.insert(0, transform)
+
+                if isinstance(next_reg, int):
+                    # Apply all transforms in reverse order of which we discovered them: We now forward propagate
+                    # the actual value!
+                    return functools.reduce(
+                        lambda acc, fn: fn(acc), transforms, next_reg
+                    )
```

Comment:
nit: I thought you could use `reduce(operator.call, transforms, next_reg)` but it does not match the order of argument. This is a bit more readable.

```
val = next_reg
for xfrm in tranforms:
    val = xfrm(next_reg)
return val
```

I don't have strong opinions though; feel free to ignore.